### PR TITLE
Add Gubernator, a GAE proxy for displaying test results stored on GCS

### PR DESCRIPTION
**Pull request by rmmh**
Currently running at http://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/kubernetes-soak-continuous-e2e-gce/5043

@spxtr @fejta

### CODE REVIEW

File: gubernator/README.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+Gubernator is a frontend for displaying Kubernetes test results stored in GCS.
+
+It runs on Google App Engine, and parses JSON and junit.xml results for display.
+
+An example URL is:
+
+	http://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/kubernetes-soak-continuous-e2e-gce/5043
```
Comment (spxtr) [inline gubernator/README.md]: I think this link dies in 90 days.

**Comment by ixdy** [comment]
Something to think about: I'd like to make http://ci-test.k8s.io and http://pr-test.k8s.io redirect to this, though we probably need bucket listing to work before we do that.

I'm also not entirely sure how we'd want to have ci-test.k8s.io and pr-test.k8s.io work in a world where not all results live in gs://kubernetes-jenkins. (Currently these redirectors assume that bucket name.)

cc @kubernetes/sig-testing

File: gubernator/main.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+app = webapp2.WSGIApplication([
+    (r'/build/([-\w]+)/([-\w/]*/)?([-\w]+)/(\d+)/?', BuildHandler),
+    # webapp2.Route('/upload', gcs_upload.Upload, 'upload'),
+], debug=True)
+
+if os.environ.get('SERVER_SOFTWARE','').startswith('Development'):
+    # inject some test data so there's a page with some content
+    import tarfile
+    tf = tarfile.open('kube_results.tar.gz')
```
Comment (fejta) [inline gubernator/main.py]: Can you ask around and see if there are policies about storing binary data in our repo? I'm not sure the answer to this question (but neat way to speed up local dev btw!)

File: gubernator/main.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        started, finished, failures = details
+        commit = started['version'].split('+')[1]
+        template = JINJA_ENVIRONMENT.get_template('build.html')
+        self.response.write(template.render(dict(
+            build_dir=build_dir, job=job, build=build, commit=commit,
+            started=started, finished=finished, failures=failures)))
+
+app = webapp2.WSGIApplication([
+    (r'/build/([-\w]+)/([-\w/]*/)?([-\w]+)/(\d+)/?', BuildHandler),
+    # webapp2.Route('/upload', gcs_upload.Upload, 'upload'),
```
Comment (fejta) [inline gubernator/main.py]: Remove?

File: gubernator/main.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    autoescape=True)
+
+
+def jinja_filter(func):
+    JINJA_ENVIRONMENT.filters[func.__name__.replace('format_', '')] = func
+    return func
+
+
+@jinja_filter
+def format_timestamp(ts):
```
Comment (fejta) [inline gubernator/main.py]: Please use longer values for function variables (func and stamp are fine but ts for timestamp is too short, same with inp which I don't even know what it is supposed to be (inpath?)

File: gubernator/main.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    t = datetime.datetime.fromtimestamp(ts)
+    return t.strftime('%F %H:%M')
+
+
+@jinja_filter
+def format_duration(seconds):
```
Comment (fejta) [inline gubernator/main.py]: Please add pydoc comments for complex functions (you can just use a single sentence summary)

File: gubernator/main.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    try:
+        with gcs.open(path) as f:
+            return f.read()
+    except gcs.errors.Error:
+        return None
+
+
+@memcache_memoize('build-details://', exp=60 * 60 * 4)
+def build_details(build_dir):
+    started = gcs_read(build_dir + '/started.json')
```
Comment (fejta) [inline gubernator/main.py]: Don't use + here

**Comment by rmmh** [comment]
Moving this to https://github.com/kubernetes/test-infra

**Pull request closed without merging**